File: dou_transit/lib/services/osrm_service.dart
import 'dart:convert';
import 'dart:io';
import 'package:http/http.dart' as http;
import 'package:latlong2/latlong.dart';

/// Service for fetching routes from the OSRM server.
class OsmService {
  static const String osrmUrl = 'https://dockerfile-dwb2.onrender.com';

  /// Get a route between two points. Returns list of LatLng points.
  static Future<List<LatLng>> getRoute(
    double startLat,
    double startLng,
    double endLat,
    double endLng, {
    List<List<double>>? waypoints,
  }) async {
    try {
      String coords = '$startLng,$startLat;$endLng,$endLat';

      if (waypoints != null && waypoints.length > 1) {
        coords = '$startLng,$startLat;';
        for (final wp in waypoints) {
          coords += '${wp[1]},${wp[0]};';
        }
        coords += '$endLng,$endLat';
      }

      final url = '$osrmUrl/route/v1/driving/$coords'
          '?overview=full&geometries=geojson&steps=true&alternatives=false';

      final response = await http
          .get(Uri.parse(url))
          .timeout(const Duration(seconds: 15));

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return _parseRoute(data);
      }
    } on SocketException {
      // Return straight line if offline
    } catch (e) {
      // Return straight line for any error
    }

    // Fallback: straight line
    return _straightLine(startLat, startLng, endLat, endLng);
  }

  static List<LatLng> _parseRoute(Map<String, dynamic> data) {
    final coords = <double>[];
    try {
      final routes = data['routes'] as List;
      if (routes.isEmpty) return [];

      final geometry = routes[0]['geometry'];
      if (geometry is Map) {
        final coordinates = geometry['coordinates'] as List;
        for (final coord in coordinates) {
          coords.add(coord[1].toDouble()); // lat
          coords.add(coord[0].toDouble()); // lng
        }
      }
    } catch (_) {
      return [];
    }

    final points = <LatLng>[];
    for (int i = 0; i < coords.length; i += 2) {
      points.add(LatLng(coords[i], coords[i + 1]));
    }
    return points;
  }

  static List<LatLng> _straightLine(double startLat, double startLng, double endLat, double endLng) {
    final points = <LatLng>[];
    final steps = 10;
    for (int i = 0; i <= steps; i++) {
      final t = i / steps;
      points.add(LatLng(
        startLat + (endLat - startLat) * t,
        startLng + (endLng - startLng) * t,
      ));
    }
    return points;
  }

  /// Get a route summary (distance in meters, duration in seconds).
  static Future<Map<String, double>> getRouteSummary(
    double startLat,
    double startLng,
    double endLat,
    double endLng,
  ) async {
    try {
      final url = '$osrmUrl/route/v1/driving/$startLng,$startLat;$endLng,$endLat'
          '?overview=false';
      final response = await http.get(Uri.parse(url)).timeout(const Duration(seconds: 15));

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        final route = (data['routes'] as List).first;
        return {
          'distance': (route['distance'] as num).toDouble(),
          'duration': (route['duration'] as num).toDouble(),
        };
      }
    } catch (_) {}

    return {'distance': 0, 'duration': 0};
  }
}

File: rn_transit/src/services/osrmService.ts
/**
 * OSRM routing service — fetches driving/walking routes from the OSRM server.
 * Mirrors Flutter's OsmService.
 */

const OSRM_URL = 'https://dockerfile-dwb2.onrender.com';

export interface RoutePoint {
  lat: number;
  lng: number;
}

export interface RouteStep {
  instruction: string;
  distanceMeters: number;
  durationSeconds: number;
  turnModifier?: string;
}

/**
 * Get a route between two points. Returns an array of {lat, lng} points.
 */
export async function getRoute(
  startLat: number,
  startLng: number,
  endLat: number,
  endLng: number,
  waypoints?: RoutePoint[],
): Promise<RoutePoint[]> {
  try {
    let coords = `${startLng},${startLat};${endLng},${endLat}`;

    if (waypoints && waypoints.length > 0) {
      coords = `${startLng},${startLat};`;
      for (const wp of waypoints) {
        coords += `${wp.lng},${wp.lat};`;
      }
      coords += `${endLng},${endLat}`;
    }

    const url = `${OSRM_URL}/route/v1/driving/${coords}?overview=full&geometries=geojson&steps=true&alternatives=false`;
    const response = await fetch(url, { signal: AbortSignal.timeout(15000) });

    if (response.ok) {
      const data = await response.json();
      return parseRoute(data);
    }
  } catch {
    // fall through to straight-line fallback
  }

  return straightLine(startLat, startLng, endLat, endLng);
}

function parseRoute(data: Record<string, unknown>): RoutePoint[] {
  try {
    const routes = data.routes as Record<string, unknown>[];
    if (!routes?.length) return [];

    const geometry = routes[0].geometry as Record<string, unknown>;
    const coordinates = geometry?.coordinates as number[][];
    if (!coordinates?.length) return [];

    return coordinates.map((coord) => ({
      lat: coord[1],
      lng: coord[0],
    }));
  } catch {
    return [];
  }
}

function straightLine(
  startLat: number,
  startLng: number,
  endLat: number,
  endLng: number,
): RoutePoint[] {
  const points: RoutePoint[] = [];
  const steps = 10;
  for (let i = 0; i <= steps; i++) {
    const t = i / steps;
    points.push({
      lat: startLat + (endLat - startLat) * t,
      lng: startLng + (endLng - startLng) * t,
    });
  }
  return points;
}

/**
 * Get a route summary (distance in meters, duration in seconds).
 */
export async function getRouteSummary(
  startLat: number,
  startLng: number,
  endLat: number,
  endLng: number,
): Promise<{ distance: number; duration: number }> {
  try {
    const url = `${OSRM_URL}/route/v1/driving/${startLng},${startLat};${endLng},${endLat}?overview=false`;
    const response = await fetch(url, { signal: AbortSignal.timeout(15000) });

    if (response.ok) {
      const data = await response.json();
      const route = (data.routes as Record<string, unknown>[])?.[0];
      if (route) {
        return {
          distance: (route.distance as number) ?? 0,
          duration: (route.duration as number) ?? 0,
        };
      }
    }
  } catch {
    // fall through
  }

  return { distance: 0, duration: 0 };
}

/**
 * Fetch turn-by-turn directions for a route.
 */
export async function getRouteSteps(
  startLat: number,
  startLng: number,
  endLat: number,
  endLng: number,
): Promise<RouteStep[]> {
  try {
    const url = `${OSRM_URL}/route/v1/driving/${startLng},${startLat};${endLng},${endLat}?overview=false&steps=true&alternatives=false`;
    const response = await fetch(url, { signal: AbortSignal.timeout(15000) });

    if (response.ok) {
      const data = await response.json();
      const routes = data.routes as Record<string, unknown>[];
      if (routes?.length) {
        const legs = routes[0].legs as Record<string, unknown>[];
        if (legs?.length) {
          const steps = legs[0].steps as Record<string, unknown>[];
          return steps.map((s) => {
            const maneuver = s.maneuver as Record<string, unknown> | undefined;
            return {
              instruction: (maneuver?.instruction as string) ?? 'Continue',
              distanceMeters: (s.distance as number) ?? 0,
              durationSeconds: (s.duration as number) ?? 0,
              turnModifier: maneuver?.modifier as string | undefined,
            };
          });
        }
      }
    }
  } catch {
    // fall through
  }

  return [];
}
